src/main/ts/core/Changelog.ts

```diff
@@ -0,0 +1,361 @@
+import * as commonmark from 'commonmark';
+import { DateTime } from 'luxon';
+import * as E from 'fp-ts/Either';
+import { pipe } from 'fp-ts/pipeable';
+import * as EitherUtils from '../utils/EitherUtils';
+import { parseVersionE, Version } from './Version';
+
+interface Offset {
+  start: number;
+  end: number;
+}
+
+interface Item extends Offset {
+  jira?: string;
+}
+
+interface Section extends Offset {
+  header: Offset;
+  list: Offset;
+  items: Item[];
+}
+
+type SectionNames = 'Added' | 'Improved' | 'Changed' | 'Deprecated' | 'Removed' | 'Fixed' | 'Security';
+
+interface ChangelogFragment {
+  sections: SectionNames[];
+  Added?: Section;
+  Improved?: Section;
+  Changed?: Section;
+  Deprecated?: Section;
+  Removed?: Section;
+  Fixed?: Section;
+  Security?: Section;
+}
+
+interface ReleaseMeta {
+  version: Version;
+  date: DateTime;
+}
+
+interface Release extends ChangelogFragment, Offset {
+  header: Offset;
+  meta?: ReleaseMeta;
+}
+
+export interface Changelog {
+  source: string;
+  releases: Release[];
+  preamble: Offset;
+  links: Offset;
+}
+
+interface TextLines {
+  text: string;
+  lines: number[];
+}
+
+interface Heading {
+  header: commonmark.Node;
+  preamble: commonmark.Node[];
+  subheadings: Heading[];
+}
+
+interface Top {
+  header?: undefined;
+  preamble: commonmark.Node[];
+  subheadings: Heading[];
+}
+
+const findLineStarts = (text: string): TextLines => {
```

Comment:
Is this just doing a split on CRLF | CR | LF?

Reply:
Oh, it's finding offsets?

Reply:
Do we even need to handle CR? If we're only storing the starts of the lines, then there's no difference between a line that ends with CRLF and a line that ends with LF that happens to have a CR right before the end of the line. Also, at least for our repos, we use unix line endings for everything don't we?

Reply:
Quite a lot of the world uses Windows. I remember running into some Java tooling when I was much younger that failed to handle CRLF and getting utterly confusing results. I would rather not be the cause of such software...

Reply:
Though reading your comment, apparently some early Macs used LFCR too. It's not too hard to handle them all.